arch/arm/core/elf.c

```diff
@@ -11,42 +12,265 @@
+		prel31_reloc(loc, &offset);
+	}
+
+	return ret;
+}
+
+static inline int jumps_decode(elf_word reloc_type, uint32_t loc,
+				uint32_t sym_base_addr, const char *sym_name, int32_t *offset)
+{
+	int ret;
+
+	*offset = MEM2ARMOPCODE(*(uint32_t *)loc);
+	*offset = (*offset & MASK_BRANCH_OFFSET) << SHIFT_BRANCH_OFFSET;
+	*offset = sign_extend(*offset, SHIFT_JUMPS_SIGN);
+	*offset += sym_base_addr - loc;
+	if (*offset <= JUMP_UPPER_BOUNDARY || *offset >= JUMP_LOWER_BOUNDARY) {
+		LOG_ERR("sym '%s': relocation out of range (%#x -> %#x)\n",
+			sym_name, loc, sym_base_addr);
+		ret = -ENOEXEC;
+	} else {
+		ret = 0;
+	}
+
+	return ret;
+}
+
+static inline void jumps_reloc(uint32_t loc, int32_t *offset)
+{
+	*offset >>= SHIFT_BRANCH_OFFSET;
+	*offset &= MASK_BRANCH_OFFSET;
 
-	addend <<= ARM_BL_BLX_ADDEND_SIZE;
-	addend |= upper & ARM_BL_BLX_ADDEND_MASK;
-	addend <<= ARM_BL_BLX_ADDEND_SIZE;
-	addend |= lower & ARM_BL_BLX_ADDEND_MASK;
+	*(uint32_t *)loc &= OPCODE2ARMMEM(MASK_BRANCH_COND|MASK_BRANCH_101|MASK_BRANCH_L);
+	*(uint32_t *)loc |= OPCODE2ARMMEM(*offset);
+}
+
+static int jumps_handler(elf_word reloc_type, uint32_t loc,
+				uint32_t sym_base_addr, const char *sym_name)
+{
+	int ret;
+	int32_t offset;
+
+	ret = jumps_decode(reloc_type, loc, sym_base_addr, sym_name, &offset);
+	if (!ret) {
+		jumps_reloc(loc, &offset);
+	}
+
+	return ret;
+}
+
+static void movs_handler(elf_word reloc_type, uint32_t loc,
+			uint32_t sym_base_addr, const char *sym_name)
+{
+	int32_t offset;
+	uint32_t tmp;
+
+	offset = tmp = MEM2ARMOPCODE(*(uint32_t *)loc);
+	offset = ((offset & MASK_MOV_RN) >> SHIFT_MOV_RN) | (offset & MASK_MOV_OPERAND2);
+	offset = sign_extend(offset, SHIFT_MOVS_SIGN);
 
-	return lower & ARM_BL_BLX_LOWER_T1T2_BIT ? addend << 1 : addend << 2;
+	offset += sym_base_addr;
+	if (reloc_type == R_ARM_MOVT_PREL || reloc_type == R_ARM_MOVW_PREL_NC) {
+		offset -= loc;
+	}
+	if (reloc_type == R_ARM_MOVT_ABS || reloc_type == R_ARM_MOVT_PREL) {
+		offset >>= 16;
+	}
+
+	tmp &= (MASK_MOV_COND | MASK_MOV_00 | MASK_MOV_I | MASK_MOV_OPCODE | MASK_MOV_RD);
+	tmp |= ((offset & MASK_MOV_RD) << SHIFT_MOV_RD) | (offset & MASK_MOV_OPERAND2);
+
+	*(uint32_t *)loc = OPCODE2ARMMEM(tmp);
 }
 
-static void arm_bl_blx_encode_addend(uintptr_t opaddr, int32_t addend)
+static inline int thm_jumps_decode(elf_word reloc_type, uint32_t loc,
+			uint32_t sym_base_addr, const char *sym_name, int32_t *offset,
+			uint32_t *upper, uint32_t *lower)
 {
-	uint16_t upper = *((uint16_t *)opaddr);
-	uint16_t lower = *(((uint16_t *)opaddr) + 1);
+	int ret;
+	uint32_t j_one, j_two, sign;
+
+	*upper = MEM2THM16OPCODE(*(uint16_t *)loc);
+	*lower = MEM2THM16OPCODE(*(uint16_t *)(loc + 2));
 
-	addend = upper & ARM_BL_BLX_UPPER_S_BIT ? addend >> 1 : addend >> 2;
+	/* sign is bit10 */
+	sign = (*upper >> BIT_THM_BW_S) & 1;
+	j_one = (*lower >> BIT_THM_BL_J1) & 1;
+	j_two = (*lower >> BIT_THM_BL_J2) & 1;
+	*offset = (sign << SHIFT_THM_JUMPS_SIGN) | ((~(j_one ^ sign) & 1) << SHIFT_THM_BL_J1) |
```

Comment:
Nit for ease of reading:
```suggestion
	*offset = (sign << SHIFT_THM_JUMPS_SIGN) |
				((~(j_one ^ sign) & 1) << SHIFT_THM_BL_J1) |
```